#763 wiki update for removal step (#833)
Azure/ResourceModules: docs/wiki/PipelineRemovalAction.md

> # Removal action

Reviewer: Placeholder for review

Reviewer 2: Love it :D

This section describes how the removal of resources deployed by a module is performed and how to modify the default behaviour if a specific module or resource type needs it.

---

### _Navigation_

- [Overview](#Overview)
- [How it works](#how-it-works)
- [Create a specialized removal procedure](#create-a-specialized-removal-procedure)

---

> ## Overview

Reviewer 2: ```suggestion
# Overview
```

Reviewer 2: For consistency

Author: I will get this change for the sake of alignment between the different pages, but having multiple 1st level title is quite ugly, especially if we have a clear hierarchical relationship between sections (e.g. the "overview" of the "removal action")

Reviewer 2: I do agree, BUT, let me at least lay down why:
- GitHub unfortunately does not support page headers - all has to be in text
- The title of the page naturally has to be `#` as otherwise it will not show up prominently as the page title
- All subsequent headers could be `##` (and that's how it was initially, actually). However, this essentially wasts one `#` and this becomes a problem for pages that use e.g. 4 indents as that's when it stops. This became a problem for several wiki pages, so we refactored it to essentially have the page title on the same level as the actual main chapter headers

The Removal action is triggered after the deployment completes. This is used for several reasons:
- Make sure to keep the validation subscription cost as low as possible.
- Enable testing of the full module deployment at every run.

The default removal procedure works fine for most of the modules created so far, so it's likely you won't have to change anything to make the module you're editing to be removed correctly after deployment.

> ## How it works

Reviewer 2: ```suggestion
# How it works
```

Reviewer 2: For consistency

1. Filtering out from the list any resource used as dependencies for different modules (e.g. the commonly used Log Analytics workspace)
1. Moving specific resource types to the top of the list (if a certain order is required). For example `vWAN` requires its `Virtual Hubs` to be removed first, even though they are no child-resources.

After a resource is removed (this happens after each resource in the list), the script will execute, if defined, a **post removal operation**. This can be used for those resource types that requires a post processing, like purging a soft-deleted key vault.

The procedure is initiated by the script `/utilities/pipelines/resourceRemoval/Initialize-DeploymentRemoval.ps1`, run during deployment by:
- (Azure DevOps) `/.azuredevops/pipelineTemplates/module.jobs.deploy.yml`
- (GitHub) `/.github/actions/templates/validateModuleDeployment/action.yml`

It uses several helper scripts that can be found in the `/utilities/pipelines/resourceRemoval/helper` folder

> ## Create a specialized removal procedure

Reviewer 2: ```suggestion
# Create a specialized removal procedure
```

Reviewer 2: For consistency

You can define a custom removal procedure by:

> 1. influencing the **sequence** in which resource are removed by prioritizing specific resource types (_Virtual WAN_ and _Automation Account_ are examples of modules that uses this method)

Reviewer 2: Good thing that you added the 3rd option, however, as per the original comment I'd suggest to move the example into a intended bullet for each path to allow us to explain 'why' e.g. an RSV would need a custom removal action. Just to say it needs one does not yet provide the full picture.

As a reference:
#1 is required for example for _Virtual WAN_ as it requires its connected resources to be removed first
#2 is required for example for _Recovery Services Vaults_ as they may contain protected items
#3 is required for example for _Key Vaults_ to purge them if they are soft-deleted

Author: Hmmm... I agree simply mentioning the resource type doesn't provide any additional value.
But I don't know... I would even remove the example, that are described above in the "how it works" section.
I prefer to have a clean list and don't repeat anything.
Maybe it's a personal thing, but for every additional second it takes to find what I need, chances I will abandon the reading dramatically increase

Reviewer 2: But would the further indent not make clear that what ever comes after is an example one can skip/read over? I see your point and there is no clear answer.
Forthunatly, we have @eriqua to save the day ;)

Reviewer: Not sure if I got the problem here but if the issue is readablity we may want to put the example in a note. For instance:

```suggestion
1. influencing the **order** in which resources are removed by prioritizing specific resource types 
    > **Example** Removing a _Virtual WAN_ resource requires related resources to be deleted in a specific order
```

> 1. define a **custom removal action** to remove a resource of a _specific resource type_ (_Recovery Services Vault_ is an example of resource type that defines it)

Reviewer: For consistency 
```suggestion
1. defining a **custom removal action** to remove a resource of a _specific resource type_ 
    > **Example** A _Recovery Services Vault_ resource requires some protected items to be identified and removed beforehand
```

> 1. define a custom **post-removal action** to be run after removing a resource of a _specific resource type_ (_Key Vaults_ uses this to purge)

Reviewer: ```suggestion
1. defining a custom **post-removal action** to be run after removing a resource of a _specific resource type_
    > **Example** A _Key Vault_ resource needs to be purged when soft deletion is enforced
```

Those methods can be combined independently.

> **Important**: _custom_ and _post-removal_ actions will be executed when a resource of the type you specify is removed **regardless** of which deployment triggered the deployment. Make sure you do not assume the resource is in a particular state defined by your module.

> To modify the resource types removal **sequence**:

Reviewer: ```suggestion
To modify the resource types removal **order**:
```
